### PR TITLE
Update 'Client' to hold stack of connection/batch/transaction objects.

### PR DESCRIPTION
**Pull request by tseaver**
See #944.

### CODE REVIEW

**Comment by dhermes** [comment]
1. The connection stack is meant to be like the `_LocalStack`? Can we give it a different name since it will presumably also hold transactions (and other non-connections).
1. This stack isn't threadsafe. Is that relevant?
1. Is the rest of the change coming soon?

**Comment by tseaver** [comment]
>   The connection stack is meant to be like the _LocalStack? Can we give it a different name since it will presumably also hold transactions (and other non-connections).

They must all be implement the "connection" API.  I can't think of a better name.

>  This stack isn't threadsafe. Is that relevant?

`list.append` and `list.pop` are atomic operations, which means that the list won't be corrupted by separate threads calling `_push_connection`.  Threads using different batch / transaction objects with the same client could end up mangling one another:  if it matters, we could just use the `gcloud._helpers._LocalStack` instead.

>  Is the rest of the change coming soon?

I think so.  I threw out a much bigger changeset and started over, but I think this is simpler / clearer.

**Comment by dhermes** [comment]
But `Transaction` doesn't implement the same API as `Connection`. Am I missing something? Are there other things that will go on this stack?

Yeah the thread-safety issue is not resolved by this, can you file an issue so that we follow up? `_LocalStack` makes the stack local to only 1 thread?

**Comment by tseaver** [comment]
@dhermes 82ee82c should address both your issues:  it makes the stack hold only batch / transaction objects, and uses a `_LocalStack` for it (so that separate threads don't collide).

File: gcloud/datastore/test_client.py [hunk 1]
```diff
@@ -74,6 +75,25 @@ def test_ctor_w_explicit_inputs(self):
         self.assertEqual(client.dataset_id, OTHER)
         self.assertEqual(client.namespace, NAMESPACE)
         self.assertTrue(client.connection is conn)
+        self.assertTrue(client.current_batch is None)
+        self.assertEqual(list(client._batch_stack), [])
+
+    def test__push_connection_and__pop_connection(self):
```
Comment (dhermes) [inline gcloud/datastore/test_client.py]: Small nit: it seems the name of this test should have changed.

File: gcloud/datastore/client.py [hunk 2]
```diff
@@ -127,8 +128,42 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
         if connection is None:
             connection = get_connection()
         self.connection = connection
+        self._batch_stack = _LocalStack()
```
Comment (dhermes) [inline gcloud/datastore/client.py]: Is it expensive to create an instance? Is it worth holding off (via a lazy-loaded property) until it is used?

**Comment by dhermes** [comment]
LGTM other than small nit. I think the answer to my "is it expensive" question is "no", but may you can educate me.

**Comment by tseaver** [comment]
I have a fix for the test name on the next PR.

ISTM that the expense of creating a thread-local object isn't really an issue for our `Client`, which will typically be constructed only at startup:  the q_BATCHES` global is going away, so it will be net zero for the typical case.

**Comment by dhermes** [comment]
Good call on net zero.

**Pull request merged**